### PR TITLE
OPN-1233 Implement Multisig voting module and integrate it with supra governance

### PR DESCRIPTION
**Pull request by nizam-supraoracles**

### CODE REVIEW

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 1]
```diff
@@ -166,6 +171,74 @@ module supra_framework::supra_governance {
...
+    /// Event emitted when the governance configs are updated.
+    struct SupraUpdateConfig has drop, store {
+        voting_duration_secs: u64,
+        min_voting_threshold: u64,
+        voters: vector<address>,
+    }
+
+    /// Used to track which execution script hashes have been approved by governance.
+    /// This is required to bypass cases where the execution scripts exceed the size limit imposed by mempool.
+    struct SupraApprovedExecutionHashes has key {
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Use the existing `ApprovedExecutionHashes` since this is accessed by Rust layer. As it is, this is not exposed via public interface.

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 2]
```diff
@@ -0,0 +1,1327 @@
...
+        expiration_secs: u64,
+
+        /// Number of votes for each outcome.
+        yes_votes: u64,
+        no_votes: u64,
+
+        /// list of voters
+        voters: vector<address>,
+
+        /// list of voters who completed there vote
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move]:
```suggestion
        /// list of voters who completed their vote
```

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 3]
```diff
@@ -0,0 +1,1327 @@
...
+        voters: vector<address>,
+        expiration_secs: u64,
+        metadata: SimpleMap<String, vector<u8>>,
+        is_multi_step_proposal: bool,
+    ): u64 acquires VotingForum {
+
+        // Make sure the execution script's hash is not empty.
+        assert!(vector::length(&execution_hash) > 0, error::invalid_argument(EPROPOSAL_EMPTY_EXECUTION_HASH));
+
+        // Make sure voters length must grether than or equal to vote threshold
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move]:
```suggestion
        // Make sure voters length must greater or equal  to vote threshold
```

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 4]
```diff
@@ -0,0 +1,1327 @@
...
+        assert!(vector::length(&voters) >= min_vote_threshold, error::invalid_argument(ETHRESHOLD_EXCEEDS_VOTERS));
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move:297]: Also, assertion is required that `min_vote_threshold>1` (must require more than one approval)

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 5]
```diff
@@ -0,0 +1,1327 @@
...
+    /// Vote on the given proposal.
+    ///
+    /// @param _proof Required so only the governance module that defines ProposalType can initiate voting.
+    /// @param voter signer of the voter.
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param should_pass Whether the votes are for yes or no.
+    public fun vote<ProposalType: store>(
+        _proof: &ProposalType,
+        voter: &signer,
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move]: Is `_proof : &ProposalType` also the first argument in `voting` module? Typically `signer` is the first argument. Please check.

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 6]
```diff
@@ -0,0 +1,1327 @@
...
+            error::invalid_state(EMULTI_STEP_PROPOSAL_IN_EXECUTION));
+
+        if (table::contains(&proposal.voted_records, voter_address)) {
+            let voted_record = table::borrow_mut(&mut proposal.voted_records, voter_address);
+            assert!(voted_record != &should_pass, error::already_exists(EVOTE_DUPLICATE_VOTE));
+            *voted_record = should_pass;
+
+            if (should_pass) {
+                proposal.yes_votes = proposal.yes_votes + 1;
+                proposal.no_votes = proposal.no_votes - 1;
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move:396]: Why is this needed? If `should_pass` is `true` then you only increase `yes_votes`, subtracting from `no_votes` results in subtraction underflow. Not needed.

Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move:401]:
```suggestion
 // Since the entry existed in `voted_records`, the vote flip should be reflected by this subtraction
                proposal.no_votes = proposal.no_votes - 1;
```

File: aptos-move/framework/supra-framework/sources/multisig_voting.move [hunk 7]
```diff
@@ -0,0 +1,1327 @@
...
+            } else {
+                proposal.yes_votes = proposal.yes_votes - 1;
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/multisig_voting.move:404]:
```suggestion
// Since the voter already voted and flipping the vote here, this subtraction is needed.
                proposal.yes_votes = proposal.yes_votes - 1;
```

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 8]
```diff
@@ -217,6 +290,29 @@ module supra_framework::supra_governance {
         })
     }
 
+    fun supra_initialize(
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Use `initialize` name only so that vm erroneously doesn't call the incorrect version. Remove the other/older version of the function.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 9]
```diff
@@ -217,6 +290,29 @@ module supra_framework::supra_governance {
...
+            voting_duration_secs,
+            min_voting_threshold,
+            voters,
+        });
+        move_to(supra_framework, SupraGovernanceEvents {
+            create_proposal_events: account::new_event_handle<SupraCreateProposalEvent>(supra_framework),
+            update_config_events: account::new_event_handle<SupraUpdateConfigEvent>(supra_framework),
+            vote_events: account::new_event_handle<SupraVoteEvent>(supra_framework),
+        });
+        move_to(supra_framework, SupraApprovedExecutionHashes {
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Again, just use original `ApprovedExecutionHashes` and not create another `struct` otherwise this would result in Rust layer panic. Rust layer depends on this particular `struct` with this particular name.

File: aptos-move/framework/supra-framework/sources/supra_governance.move (hunk 1 above)
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Do not create a new struct, use `ApprovedExecutionHashes` only since Rust layer depends on it.

File: aptos-move/framework/supra-framework/sources/supra_governance.move (hunk 8 above)
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Use `initialize` as a name since VM is going to call `initialize`. This is a private function so should not create errors. Same with events, I don't think creating new set of events with `Supra` prefix is necessary.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 10]
```diff
@@ -252,6 +348,39 @@ module supra_framework::supra_governance {
         );
     }
 
+    public fun update_supra_governance_config(
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move:353]: As much as possible see if original function names can be used.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 11]
```diff
@@ -252,6 +348,39 @@ module supra_framework::supra_governance {
...
+        supra_framework: &signer,
+        voting_duration_secs: u64,
+        min_voting_threshold: u64,
+        voters: vector<address>,
+    ) acquires SupraGovernanceConfig, SupraGovernanceEvents {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let supra_governance_config = borrow_global_mut<SupraGovernanceConfig>(@supra_framework);
+        supra_governance_config.voting_duration_secs = voting_duration_secs;
+        supra_governance_config.min_voting_threshold = min_voting_threshold;
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move:366]: `assert!(vector::length(&voter) >= min_voting_threshold && min_voting_threshold > vector::length(&voter)/2 && min_voting_threshold > 2, ERROR)`
These checks are needed during upgrade
Also `voting_duration` should not be too small so put an assertion that it is at least 1 hour.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 12]
```diff
@@ -449,6 +578,94 @@ module supra_framework::supra_governance {
         proposal_id
     }
 
+
+    /// Create a single-step proposal with the backing `stake_pool`.
+    /// @param execution_hash Required. This is the hash of the resolution script. When the proposal is resolved,
+    /// only the exact script with matching hash can be successfully executed.
+    public entry fun supra_create_proposal(
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move:590]: Can we use original function name here?

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 13]
```diff
@@ -581,10 +798,77 @@ module supra_framework::supra_governance {
...
+
+        // The voter's stake needs to be locked up at least as long as the proposal's expiration.
+        let proposal_expiration = multisig_voting::get_proposal_expiration_secs<GovernanceProposal>(
+            @supra_framework,
+            proposal_id
+        );
+        assert!(timestamp::now_seconds() >= proposal_expiration, error::invalid_argument(EPROPOSAL_IS_EXPIRE));
+
+        multisig_voting::vote<GovernanceProposal>(
+            &governance_proposal::create_empty_proposal(),
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Did you check what this `create_empty_proposal` does?

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 14]
```diff
@@ -581,10 +798,77 @@ module supra_framework::supra_governance {
...
+        }
+    }
+
     public entry fun add_approved_script_hash_script(proposal_id: u64) acquires ApprovedExecutionHashes {
         add_approved_script_hash(proposal_id)
     }
 
+    public entry fun add_supra_approved_script_hash_script(proposal_id: u64) acquires SupraApprovedExecutionHashes {
+        add_supra_approved_script_hash(proposal_id)
+    }
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Use original function. If the argument list/type is not changing, no need to rename function.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 15]
```diff
@@ -607,6 +891,28 @@ module supra_framework::supra_governance {
         }
     }
 
+    /// Add the execution script hash of a successful governance proposal to the approved list.
+    /// This is needed to bypass the mempool transaction size limit for approved governance proposal transactions that
+    /// are too large (e.g. module upgrades).
+    public fun add_supra_approved_script_hash(proposal_id: u64) acquires SupraApprovedExecutionHashes {
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: You can/should reuse original function here.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 16]
```diff
@@ -618,6 +924,17 @@ module supra_framework::supra_governance {
         get_signer(signer_address)
     }
 
+    /// Resolve a successful single-step proposal. This would fail if the proposal is not successful (not enough votes or more no
+    /// than yes).
+    public fun supra_resolve(
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move:934]: Use original function here for `resolve`.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 17]
```diff
@@ -638,6 +955,26 @@ module supra_framework::supra_governance {
         get_signer(signer_address)
     }
 
+    /// Resolve a successful multi-step proposal. This would fail if the proposal is not successful.
+    public fun resolve_supra_multi_step_proposal(
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move:964]: Same as above

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 18]
```diff
@@ -651,6 +988,19 @@ module supra_framework::supra_governance {
         };
     }
 
+    /// Remove an approved proposal's execution script hash.
+    public fun remove_supra_approved_hash(proposal_id: u64) acquires SupraApprovedExecutionHashes {
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Same as above.

File: aptos-move/framework/supra-framework/sources/supra_governance.move [hunk 19]
```diff
@@ -1381,7 +1731,10 @@ module supra_framework::supra_governance {
         min_voting_threshold: u128,
         required_proposer_stake: u64,
         voting_duration_secs: u64,
+        supra_min_voting_threshold: u64,
+        voters: vector<address>,
     ) {
         initialize(supra_framework, min_voting_threshold, required_proposer_stake, voting_duration_secs);
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/supra_governance.move]: Use `initialize` for our specific initializtion, no need for `supra_initialize`

**Review by sjoshisupra — COMMENTED** [review]
Address the comments. Add more test cases, use original functions as much as possible.

File: aptos-move/framework/supra-framework/sources/supra_governance.move (hunk 11 above)
Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/supra_governance.move:366]: Make sure these checks are present also in `initialize`

File: aptos-move/framework/supra-framework/sources/multisig_voting.move (hunk 5 above)
Comment (nizam-supraoracles) [reply aptos-move/framework/supra-framework/sources/multisig_voting.move]: Since this `vote` function is not an entry function, the order of the signer parameter does not matter. The sequence depends on how we call it, and we need to maintain the same order.

However, it would be better if the signer parameter is in the first place

File: aptos-move/framework/supra-framework/sources/supra_governance.move (hunk 13 above)
Comment (nizam-supraoracles) [reply aptos-move/framework/supra-framework/sources/supra_governance.move]: This function simply return `GovernanceProposal {}`

**Review by sjoshisupra — APPROVED** [review]
